File: Python_TRR/11.py
# Viết băng Python
# Bài 11: Thuật toán Sàng Eratosthenes.

def sieve_of_eratosthenes(n):
    """Tìm tất cả các số nguyên tố nhỏ hơn hoặc bằng n."""
    primes = [True] * (n + 1)  # Tạo mảng đánh dấu, mặc định là True
    primes[0] = primes[1] = False  # 0 và 1 không phải số nguyên tố

    p = 2
    while p * p <= n:
        if primes[p]:  # Nếu p là số nguyên tố
            for i in range(p * p, n + 1, p):  # Đánh dấu các bội số của p
                primes[i] = False
        p += 1

    # Trả về danh sách các số nguyên tố
    return [i for i in range(n + 1) if primes[i]]

print(sieve_of_eratosthenes(30))  # Ví dụ: Tìm số nguyên tố <= 30

File: C_TRR/11.c
// Viết bằng C
// Bài 11: Thuật toán Sàng Eratosthenes.

#include <stdio.h>
#include <stdbool.h>

void sieve_of_eratosthenes(int n) // n là giới hạn số nguyên tố cần tìm
{
    bool primes[n + 1]; // Tạo mảng đánh dấu với tất cả giá trị là true
    for (int i = 0; i <= n; i++)
    {
        primes[i] = true;  // Khởi tạo mảng đánh dấu True
    }
    primes[0] = primes[1] = false;  // 0 và 1 không phải số nguyên tố

    for (int p = 2; p * p <= n; p++) 
    {
        if (primes[p]) {  // Nếu p là số nguyên tố
            for (int i = p * p; i <= n; i += p) 
            {  // Đánh dấu bội số của p
                primes[i] = false;
            }
        }
    }

    printf("Các số nguyên tố <= %d là: ", n);
    for (int i = 2; i <= n; i++)  // In ra các số nguyên tố từ 2 đến n
    {
        if (primes[i])  // Nếu i là số nguyên tố
        {
            printf("%d ", i);
        }
    }
}

int main() 
{
    sieve_of_eratosthenes(30); 
    // Các số nguyên tố <= 30 là: 2 3 5 7 11 13 17 19 23 29
    return 0;
}